core/src/main/scala/org/apache/spark/status/api/v1/EventLogDownloadResource.scala

```diff
@@ -0,0 +1,101 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.spark.status.api.v1
+
+import java.io.{BufferedInputStream, FileInputStream, OutputStream}
+import javax.ws.rs.{GET, Produces}
+import javax.ws.rs.core.{Response, StreamingOutput, MediaType}
+
+import scala.collection.mutable.ArrayBuffer
+
+import org.apache.hadoop.fs.Path
+
+import org.apache.spark.{Logging, SparkConf}
+import org.apache.spark.deploy.SparkHadoopUtil
+import org.apache.spark.deploy.history.HistoryServer
+import org.apache.spark.util.Utils
+
+@Produces(Array(MediaType.APPLICATION_OCTET_STREAM))
+private[v1] class EventLogDownloadResource(
+    val uIRoot: UIRoot,
+    val appId: String,
+    val attemptId: Option[String]) extends Logging {
+  val conf = SparkHadoopUtil.get.newConfiguration(new SparkConf)
+
+  @GET
+  def getEventLogs(): Response = {
+    uIRoot match {
+      case hs: HistoryServer =>
+        var logsNotFound = false
+        val fileName: String = {
+          attemptId match {
+            case Some(id) => s"eventLogs-$appId-$id.zip"
+            case None => s"eventLogs-$appId.zip"
+          }
+        }
+        val stream = new StreamingOutput {
+          override def write(output: OutputStream): Unit = {
+            attemptId match {
+              case Some(id) =>
+                Utils.zipFilesToStream(hs.getEventLogPaths(appId, id), conf, output)
+              case None =>
+                val appInfo = hs.getApplicationInfoList.find(_.id == appId)
+                appInfo match {
+                  case Some(info) =>
+                    val attempts = info.attempts
+                    val files = new ArrayBuffer[Path]
+                    attempts.foreach { attempt =>
+                      attempt.attemptId.foreach { attemptId =>
+                        logInfo(s"Attempt found: ${attemptId}")
+                        files ++= hs.getEventLogPaths(appId, attemptId)
+                      }
+                    }
+                    if (files.nonEmpty) {
+                      Utils.zipFilesToStream(files, conf, output)
+                    }
+                  case None => logsNotFound = true
+                }
+            }
+            output.flush()
+          }
+        }
+        if (logsNotFound) {
+          Response.serverError()
+            .entity(s"Event logs are not available for app: $appId.")
+            .status(Response.Status.SERVICE_UNAVAILABLE)
+            .build()
+        } else {
+          Response.ok(stream)
+            .header("Content-Disposition", s"attachment; filename=$fileName")
+            .header("Content-Type", MediaType.APPLICATION_OCTET_STREAM)
+            .build()
+        }
+      case _ =>
+        Response.serverError()
+          .entity("History Server is not running - cannot return event logs.")
```

Comment:
nit: "Event logs are only available through the history server."

Is that true, though? I'm sure `Master` or even `SparkUI` could be modified to serve the logs, even if partial. I'd extend `UIRoot` to have the `getEventLogPaths()` method and, if you don't want to implement it for the others now, just throw an appropriate exception in the default implementation.